### PR TITLE
feat: [evmos] implement evmos inflation module

### PR DESCRIPTION
**Pull request by huichiaotsou**
## Description

Closes: #XXXX

---

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [ ] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] added `!` to the type prefix if API or client breaking change
- [ ] targeted the correct branch
- [ ] provided a link to the relevant issue or specification
- [ ] added a changelog entry to `CHANGELOG.md`
- [ ] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [ ] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed `!` in the type prefix if API or client breaking change
- [ ] confirmed all author checklist items have been addressed
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

### CODE REVIEW

File: database/evmos_inflation.go [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+	inflationRate := data.InflationRate.Mul(sdk.NewDecWithPrec(1, 2)).String()
+
+	_, err := db.Sql.Exec(stmt,
+		pq.Array(dbtypes.NewDbDecCoins(data.CirculatingSupply)),
+		pq.Array(dbtypes.NewDbDecCoins(data.EpochMintProvision)),
+		inflationRate, data.InflationPeriod, data.SkippedEpochs, data.Height,
+	)
+
+	if err != nil {
+		return fmt.Errorf("error while storing evmos inflation params: %s", err)
```
Comment (MonikaCat) [inline database/evmos_inflation.go]:
```suggestion
		return fmt.Errorf("error while storing evmos inflation data: %s", err)
```

File: modules/inflation/handle_genesis.go [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+// HandleGenesis implements modules.Module
+func (m *Module) HandleGenesis(doc *tmtypes.GenesisDoc, appState map[string]json.RawMessage) error {
+	log.Debug().Str("module", "inflation").Msg("parsing genesis")
+
+	// Read the genesis state
+	var genState inflationtypes.GenesisState
+	err := m.cdc.UnmarshalJSON(appState[inflationtypes.ModuleName], &genState)
+	if err != nil {
+		return fmt.Errorf("error while reading mint genesis data: %s", err)
```
Comment (MonikaCat) [inline modules/inflation/handle_genesis.go]:
```suggestion
		return fmt.Errorf("error while reading evmos inflation genesis data: %s", err)
```

File: modules/inflation/handle_periodic_operations.go [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
+// RegisterPeriodicOperations implements modules.PeriodicOperationsModule
+func (m *Module) RegisterPeriodicOperations(scheduler *gocron.Scheduler) error {
+	log.Debug().Str("module", "inflation").Msg("setting up periodic tasks")
+
+	// Setup a cron job to run every midnight
+	if _, err := scheduler.Every(10).Second().Do(func() {
```
Comment (MonikaCat) [inline modules/inflation/handle_periodic_operations.go]:
```suggestion
	if _, err := scheduler.Every(1).Day().At("00:00").Do(func() {
```

File: modules/inflation/handle_periodic_operations.go [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+		utils.WatchMethod(m.updateInflationData)
+	}); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// updateInflation fetches from the REST APIs the latest value for the
```
Comment (MonikaCat) [inline modules/inflation/handle_periodic_operations.go]:
```suggestion
// updateInflationData fetches from the REST APIs the latest value for the
```

File: modules/inflation/module.go [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+	inflationsource "github.com/forbole/bdjuno/v3/modules/inflation/source"
+)
+
+var (
+	_ modules.Module                   = &Module{}
+	_ modules.GenesisModule            = &Module{}
+	_ modules.PeriodicOperationsModule = &Module{}
+)
+
+// Module represent database/mint module
```
Comment (MonikaCat) [inline modules/inflation/module.go]:
```suggestion
// Module represent database/inflation module
```

File: modules/inflation/source/local/source.go [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+	"github.com/forbole/juno/v3/node/local"
+
+	inflationsource "github.com/forbole/bdjuno/v3/modules/inflation/source"
+)
+
+var (
+	_ inflationsource.Source = &Source{}
+)
+
+// Source implements mintsource.Source using a local node
```
Comment (MonikaCat) [inline modules/inflation/source/local/source.go]:
```suggestion
// Source implements inflationsource.Source using a local node
```

File: modules/inflation/source/local/source.go [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+
+// NewSource returns a new Source instace
+func NewSource(source *local.Source, querier inflationtypes.QueryServer) *Source {
+	return &Source{
+		Source:  source,
+		querier: querier,
+	}
+}
+
+// Params implements mintsource.Source
```
Comment (MonikaCat) [inline modules/inflation/source/local/source.go]:
```suggestion
// Params implements inflationsource.Source
```

File: modules/inflation/source/local/source.go [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+
+	res, err := s.querier.Params(sdk.WrapSDKContext(ctx), &inflationtypes.QueryParamsRequest{})
+	if err != nil {
+		return inflationtypes.Params{}, nil
+	}
+
+	return res.Params, nil
+}
+
+// CirculatingSupply implements mintsource.Source
```
Comment (MonikaCat) [inline modules/inflation/source/local/source.go]:
```suggestion
// CirculatingSupply implements inflationsource.Source
```

File: modules/inflation/source/local/source.go [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+
+	res, err := s.querier.CirculatingSupply(sdk.WrapSDKContext(ctx), &inflationtypes.QueryCirculatingSupplyRequest{})
+	if err != nil {
+		return sdk.DecCoin{}, err
+	}
+
+	return res.CirculatingSupply, nil
+}
+
+// EpochMintProvision implements mintsource.Source
```
Comment (MonikaCat) [inline modules/inflation/source/local/source.go]:
```suggestion
// EpochMintProvision implements inflationsource.Source
```

File: modules/inflation/source/local/source.go [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+
+	res, err := s.querier.EpochMintProvision(sdk.WrapSDKContext(ctx), &inflationtypes.QueryEpochMintProvisionRequest{})
+	if err != nil {
+		return sdk.DecCoin{}, err
+	}
+
+	return res.EpochMintProvision, nil
+}
+
+// InflationRate implements mintsource.Source
```
Comment (MonikaCat) [inline modules/inflation/source/local/source.go]:
```suggestion
// InflationRate implements inflationsource.Source
```

File: modules/inflation/source/local/source.go [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+
+	res, err := s.querier.InflationRate(sdk.WrapSDKContext(ctx), &inflationtypes.QueryInflationRateRequest{})
+	if err != nil {
+		return sdk.Dec{}, err
+	}
+
+	return res.InflationRate, nil
+}
+
+// InflationPeriod implements mintsource.Source
```
Comment (MonikaCat) [inline modules/inflation/source/local/source.go]:
```suggestion
// InflationPeriod implements inflationsource.Source
```

File: modules/inflation/source/local/source.go [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+
+	res, err := s.querier.Period(sdk.WrapSDKContext(ctx), &inflationtypes.QueryPeriodRequest{})
+	if err != nil {
+		return 0, err
+	}
+
+	return res.Period, nil
+}
+
+// SkippedEpochs implements mintsource.Source
```
Comment (MonikaCat) [inline modules/inflation/source/local/source.go]:
```suggestion
// SkippedEpochs implements inflationsource.Source
```

File: modules/inflation/source/remote/source.go [hunk 13]
```diff
@@ -0,0 +1,87 @@
...
+	"github.com/forbole/juno/v3/node/remote"
+
+	inflationsource "github.com/forbole/bdjuno/v3/modules/inflation/source"
+)
+
+var (
+	_ inflationsource.Source = &Source{}
+)
+
+// Source implements mintsource.Source using a remote node
```
Comment (MonikaCat) [inline modules/inflation/source/remote/source.go]:
```suggestion
// Source implements inflationsource.Source using a remote node
```

File: modules/inflation/source/remote/source.go [hunk 14]
```diff
@@ -0,0 +1,87 @@
...
+
+// NewSource returns a new Source instance
+func NewSource(source *remote.Source, querier inflationtypes.QueryClient) *Source {
+	return &Source{
+		Source:  source,
+		querier: querier,
+	}
+}
+
+// Params implements mintsource.Source
```
Comment (MonikaCat) [inline modules/inflation/source/remote/source.go]:
```suggestion
// Params implements inflationsource.Source
```

File: modules/inflation/source/remote/source.go [hunk 15]
```diff
@@ -0,0 +1,87 @@
...
+func (s Source) Params(height int64) (inflationtypes.Params, error) {
+	res, err := s.querier.Params(remote.GetHeightRequestContext(s.Ctx, height), &inflationtypes.QueryParamsRequest{})
+	if err != nil {
+		return inflationtypes.Params{}, nil
+	}
+
+	return res.Params, nil
+}
+
+// CirculatingSupply implements mintsource.Source
```
Comment (MonikaCat) [inline modules/inflation/source/remote/source.go]:
```suggestion
// CirculatingSupply implements inflationsource.Source
```

File: modules/inflation/source/remote/source.go [hunk 16]
```diff
@@ -0,0 +1,87 @@
...
+func (s Source) CirculatingSupply(height int64) (sdk.DecCoin, error) {
+	res, err := s.querier.CirculatingSupply(remote.GetHeightRequestContext(s.Ctx, height), &inflationtypes.QueryCirculatingSupplyRequest{})
+	if err != nil {
+		return sdk.DecCoin{}, err
+	}
+
+	return res.CirculatingSupply, nil
+}
+
+// EpochMintProvision implements mintsource.Source
```
Comment (MonikaCat) [inline modules/inflation/source/remote/source.go]:
```suggestion
// EpochMintProvision implements inflationsource.Source
```

File: modules/inflation/source/remote/source.go [hunk 17]
```diff
@@ -0,0 +1,87 @@
...
+func (s Source) EpochMintProvision(height int64) (sdk.DecCoin, error) {
+	res, err := s.querier.EpochMintProvision(remote.GetHeightRequestContext(s.Ctx, height), &inflationtypes.QueryEpochMintProvisionRequest{})
+	if err != nil {
+		return sdk.DecCoin{}, err
+	}
+
+	return res.EpochMintProvision, nil
+}
+
+// InflationRate implements mintsource.Source
```
Comment (MonikaCat) [inline modules/inflation/source/remote/source.go]:
```suggestion
// InflationRate implements inflationsource.Source
```

File: modules/inflation/source/remote/source.go [hunk 18]
```diff
@@ -0,0 +1,87 @@
...
+func (s Source) InflationRate(height int64) (sdk.Dec, error) {
+	res, err := s.querier.InflationRate(remote.GetHeightRequestContext(s.Ctx, height), &inflationtypes.QueryInflationRateRequest{})
+	if err != nil {
+		return sdk.Dec{}, err
+	}
+
+	return res.InflationRate, nil
+}
+
+// InflationPeriod implements mintsource.Source
```
Comment (MonikaCat) [inline modules/inflation/source/remote/source.go]:
```suggestion
// InflationPeriod implements inflationsource.Source
```

File: modules/inflation/source/remote/source.go [hunk 19]
```diff
@@ -0,0 +1,87 @@
...
+func (s Source) InflationPeriod(height int64) (uint64, error) {
+	res, err := s.querier.Period(remote.GetHeightRequestContext(s.Ctx, height), &inflationtypes.QueryPeriodRequest{})
+	if err != nil {
+		return 0, err
+	}
+
+	return res.Period, nil
+}
+
+// SkippedEpochs implements mintsource.Source
```
Comment (MonikaCat) [inline modules/inflation/source/remote/source.go]:
```suggestion
// SkippedEpochs implements inflationsource.Source
```

File: modules/inflation/utils_params.go [hunk 20]
```diff
@@ -0,0 +1,23 @@
...
+	"fmt"
+
+	"github.com/rs/zerolog/log"
+
+	"github.com/forbole/bdjuno/v3/types"
+)
+
+// UpdateParams gets the updated params and stores them inside the database
+func (m *Module) UpdateParams(height int64) error {
+	log.Debug().Str("module", "mint").Int64("height", height).
```
Comment (MonikaCat) [inline modules/inflation/utils_params.go]:
```suggestion
	log.Debug().Str("module", "inflation").Int64("height", height).
```

File: types/evmos_inflation.go [hunk 21]
```diff
@@ -0,0 +1,45 @@
...
+	inflationtypes "github.com/evmos/evmos/v6/x/inflation/types"
+)
+
+// EvmosInflationParams represents the parameters of the evmos x/inflation module
+type EvmosInflationParams struct {
+	inflationtypes.Params
+	Height int64
+}
+
+// NewInflationParams allows to build a new InflationParams instance
```
Comment (MonikaCat) [inline types/evmos_inflation.go]:
```suggestion
// NewEvmosInflationParams allows to build a new EvmosInflationParams instance
```

File: types/evmos_inflation.go [hunk 22]
```diff
@@ -0,0 +1,45 @@
...
+type EvmosInflationData struct {
+	CirculatingSupply  sdk.DecCoins
+	EpochMintProvision sdk.DecCoins
+	InflationRate      sdk.Dec
+	InflationPeriod    uint64
+	SkippedEpochs      uint64
+	Height             int64
+}
+
+// NewEvmosInflationData allows to build a new InflationData instance
```
Comment (MonikaCat) [inline types/evmos_inflation.go]:
```suggestion
// NewEvmosInflationData allows to build a new EvmosInflationData instance
```

**Review by MonikaCat — APPROVED** [review]
tested, LGTM

**Comment by MonikaCat** [comment]
@huichiaotsou oh can we move log from `CHANGELOG_EVMOS.md` back to the general `CHANGELOG.md` file?

**Review by rmontagnin — APPROVED** [review]
utACK

**Pull request merged**